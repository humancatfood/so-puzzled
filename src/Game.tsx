import { useMemo, useRef, useState, ReactElement } from'react'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import { useElementSize, coordsToId } from './utils'

import GameGrid from './Grid'
import Menu from './Menu'
import Piece from './Piece'
import Slot from './Slot'
import Stage from './Stage'



const pieceSizeRatio = 2

type GameProps = {
  imgSrc: string
}

type PieceParams = {
  id: string
  pieceWidth: number
  pieceHeight: number
  width: number
  height: number
  left: number
  top: number
}


function getPieces(width: number, height: number, pieceSizeRatio: number): Record<string, PieceParams> {

  const pieces: Record<string, PieceParams> = {}

  if (!width || !height) {
    return pieces
  }

  const desiredPieceSize = Math.min(width, height) / pieceSizeRatio

  const numCols = Math.round(width / desiredPieceSize)
  const numRows = Math.round(height / desiredPieceSize)

  const pieceWidth = width / numCols
  const pieceHeight = height / numRows


  Array(numRows).fill(0).forEach((_, y) => {
    Array(numCols).fill(0).forEach((_, x) => {
      const id = coordsToId(x, y)
      pieces[id] = {
        id,
        pieceWidth,
        pieceHeight,
        width,
        height,
        left: x * pieceWidth * -1,
        top: y * pieceHeight * -1,
      }
    })
  })
  return pieces

}

function Game({ imgSrc }: GameProps) {

  const [showHelp, setShowHelp] = useState<boolean>(false)
  const [isStarted, setStarted] = useState<boolean>(false)

  const imgRef = useRef<HTMLImageElement>(null)

  const [imgWidth, imgHeight] = useElementSize(imgRef.current)

  const onLoadImage = () => setTimeout (function () {
    setStarted(true)
  }, 2000)

  const img = useMemo(() => {
    const img = new Image(imgWidth, imgHeight)
    img.src = imgSrc
    return img
  }, [imgWidth, imgHeight, imgSrc])

  const pieces = getPieces(imgWidth, imgHeight, pieceSizeRatio)

  function renderSlot(id: string): ReactElement {
    const { width, height, pieceWidth, pieceHeight, left, top } = pieces[id]
    return (
      <Slot
        onDropPiece={id => console.log('drop in slot:', id)}
      >
        <Piece
          id={id}
          pieceWidth={pieceWidth}
          pieceHeight={pieceHeight}
          width={width}
          height={height}
          left={left}
          top={top}
          img={img}
        />
      </Slot>
    )
  }

  return (
    <>
      <Menu toggleHelp={setShowHelp} />
      <div className="game-wrapper" >
        <Stage
          onDropPiece={id => console.log(`dropped on stage: ${id}`)}
        />
        <div className="grid-wrapper">
          <img
            alt="Kitty"
            src={imgSrc}
            className={`
              base-img
              ${isStarted ? 'transparent' : ''}
              ${showHelp ? 'semi-transparent' : ''}
            `}
            ref={imgRef}
            onLoad={onLoadImage}
          />
          {isStarted && (
            <GameGrid
              width={imgWidth}
              height={imgHeight}
              pieceSizeRatio={pieceSizeRatio}
              renderSlot={renderSlot}
            />
          )}
        </div>
      </div>
    </>
  )

}



type GameWrapperProps = {
  imgSrc: string
}

export default function GameWrapper({ imgSrc }: GameWrapperProps) {
  return (
    <DndProvider backend={HTML5Backend}>
      <Game imgSrc={imgSrc}/>
    </DndProvider>
  )
}
