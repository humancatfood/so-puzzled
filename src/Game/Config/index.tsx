export { ConfigProvider, useConfig } from './Config'
export type { NoseOffsetGetter } from './noseOffsets'
