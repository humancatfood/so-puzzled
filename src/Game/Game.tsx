import { useMemo, useState, ReactElement, useEffect, useRef } from 'react'

import { Grid } from './Grid'
import { Piece } from './Piece'
import ReferenceImage from './ReferenceImage'
import { Slot } from './Slot'
import { Stage } from './Stage'
import { useGameState, getGridDimensions, getIds, coordsToId } from './logic'

type GameProps = {
  img: HTMLImageElement
  difficulty: number
}

type GameInfo = {
  pieceWidth: number
  pieceHeight: number
  width: number
  height: number
}

type PieceParams = {
  x: number
  y: number
}

function getPieces(
  numCols: number,
  numRows: number,
): Record<string, PieceParams> {
  const pieces: Record<string, PieceParams> = {}
  for (let y = 0; y < numRows; y += 1) {
    for (let x = 0; x < numCols; x += 1) {
      const id = coordsToId(x, y)
      pieces[id] = { x, y }
    }
  }
  return pieces
}

function getGameInfo(
  width: number,
  height: number,
  difficulty: number,
): GameInfo {
  const { numRows, numCols } = getGridDimensions(width, height, difficulty)

  const pieceWidth = width / numCols
  const pieceHeight = height / numRows

  return {
    pieceWidth,
    pieceHeight,
    width,
    height,
  }
}

export default function Game({ img, difficulty = 2 }: GameProps) {
  const [showHelp] = useState<boolean>(false)
  const stageRef = useRef<HTMLDivElement>(null)
  const gridRef = useRef<HTMLDivElement>(null)

  const [{ width: imgWidth, height: imgHeight }, setImageSize] = useState<{
    width: number
    height: number
  }>({ width: 0, height: 0 })

  const gameInfo = getGameInfo(imgWidth, imgHeight, difficulty)

  const { numCols, numRows } = useMemo(
    () => getGridDimensions(imgWidth, imgHeight, difficulty),
    [imgWidth, imgHeight, difficulty],
  )

  const ids = useMemo(() => getIds(numRows, numCols), [numCols, numRows])

  const pieces = useMemo(() => getPieces(numCols, numRows), [numCols, numRows])

  const {
    isSolved,
    stagePieces,
    piecesToShuffle,
    getSlotPiece,
    markPiecesToBeShuffled,
    movePieceToStage,
    movePieceToSlot,
    debug,
    shufflePieces,
  } = useGameState(ids)

  useEffect(() => {
    if (piecesToShuffle.length && stageRef.current) {
      shufflePieces(
        stageRef.current?.getBoundingClientRect().toJSON(),
        [gridRef.current?.getBoundingClientRect().toJSON()],
        gameInfo.pieceWidth,
        gameInfo.pieceHeight,
      )
    }
  }, [
    piecesToShuffle,
    shufflePieces,
    gameInfo.pieceWidth,
    gameInfo.pieceHeight,
  ])

  useEffect(() => {
    if (isSolved) {
      // markPiecesToBeShuffled(ids)
    }
  }, [isSolved, markPiecesToBeShuffled, ids])

  function renderSlot(x: number, y: number): ReactElement {
    const slotId = coordsToId(x, y)
    const pieceInSlot = getSlotPiece(slotId)
    const onDropPiece = (pieceId: string) => movePieceToSlot(pieceId, slotId)

    if (pieceInSlot) {
      const { id } = pieceInSlot
      const { x, y } = pieces[id]
      const { width, height, pieceWidth, pieceHeight } = gameInfo
      return (
        <Slot onDropPiece={onDropPiece}>
          <Piece
            id={id}
            pieceWidth={pieceWidth}
            pieceHeight={pieceHeight}
            width={width}
            height={height}
            left={x * pieceWidth * -1}
            top={y * pieceHeight * -1}
            img={img}
            offset={{ x: 0, y: 0 }}
          />
        </Slot>
      )
    } else {
      return <Slot onDropPiece={onDropPiece} />
    }
  }

  debug?.()

  return (
    <>
      <p style={{ color: 'white' }}>Solved: {isSolved.toString()}</p>
      <Stage onDropPiece={movePieceToStage} ref={stageRef}>
        <div className="grid-wrapper" ref={gridRef}>
          <ReferenceImage
            img={img}
            semiTransparent={showHelp}
            onResize={setImageSize}
          />
          <Grid
            width={imgWidth}
            height={imgHeight}
            pieceSizeRatio={difficulty}
            renderSlot={renderSlot}
          />
        </div>
        {stagePieces.map(({ id, top, left }) => {
          const { width, height, pieceWidth, pieceHeight } = gameInfo
          const { x, y } = pieces[id]
          if (top != null && left != null) {
            return (
              <Piece
                key={id}
                id={id}
                pieceWidth={pieceWidth}
                pieceHeight={pieceHeight}
                width={width}
                height={height}
                left={x * pieceWidth * -1}
                top={y * pieceHeight * -1}
                img={img}
                offset={{ x: left, y: top }}
              />
            )
          }
        })}
      </Stage>
    </>
  )
}
