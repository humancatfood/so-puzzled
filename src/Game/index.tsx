import Game from './Game'
import GameWrapper, { GameWrapperProps } from './GameWrapper'

export default function WrapperGame({
  difficulty,
  imgSrc,
}: Omit<GameWrapperProps, 'Game'>) {
  return <GameWrapper Game={Game} difficulty={difficulty} imgSrc={imgSrc} />
}
