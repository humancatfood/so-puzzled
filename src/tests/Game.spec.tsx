import '@testing-library/jest-dom'
import {
  render,
  prettyDOM,
  waitForElementToBeRemoved,
} from '@testing-library/react'
import Game from '../Game'

describe('Game', () => {
  it('renders', async () => {
    const { container, getByText, queryByText, findByTestId } = render(
      <Game imgSrc="800x500.jpg" difficulty={5} />,
    )

    const match = (snapshotName: string) => {
      expect(
        prettyDOM(container, 999999, { highlight: false }),
      ).toMatchSnapshot(snapshotName)
    }

    expect(getByText('loading')).toBeInTheDocument()
    match('loading')

    await waitForElementToBeRemoved(() => queryByText('loading'))

    await findByTestId('piece-0-0')
    match('after placing pieces')
  })
})
