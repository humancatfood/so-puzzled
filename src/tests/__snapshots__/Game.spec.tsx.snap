// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders: after placing pieces 1`] = `
"<div>
  <p
    style=\\"color: white;\\"
  >
    Solved: 
    true
  </p>
  <div
    class=\\"sc-pVTFL jFIsKS\\"
  >
    <div
      class=\\"sc-eCImPb cPhZRz\\"
    >
      <img
        class=\\"sc-furwcr lbdCTG\\"
        src=\\"800x500.jpg\\"
      />
      <div
        class=\\"sc-gsDKAQ fEJybW\\"
      >
        <div
          class=\\"sc-dkPtRN dxQedm\\"
        >
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-0-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-1-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-2-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-3-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-4-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-5-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-6-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-7-0\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class=\\"sc-dkPtRN dxQedm\\"
        >
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-0-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-1-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-2-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-3-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-4-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-5-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-6-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-7-1\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class=\\"sc-dkPtRN dxQedm\\"
        >
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-0-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-1-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-2-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-3-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-4-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-5-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-6-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-7-2\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class=\\"sc-dkPtRN dxQedm\\"
        >
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-0-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-1-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-2-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-3-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-4-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-5-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-6-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-7-3\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class=\\"sc-dkPtRN dxQedm\\"
        >
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-0-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-1-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-2-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-3-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-4-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-5-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-6-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
          <div
            class=\\"sc-hKwDye gpZIxo\\"
          >
            <div
              class=\\"sc-bdvvtL fyPoMG\\"
            >
              <div
                class=\\"sc-jRQBWg cjsshw\\"
                data-testid=\\"piece-7-4\\"
                draggable=\\"true\\"
                style=\\"width: 100px; height: 100px; top: 0px; left: 0px; cursor: grab; opacity: 1;\\"
              >
                <canvas
                  class=\\"sc-iCfMLu bEFJLT\\"
                  height=\\"160\\"
                  width=\\"160\\"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Game renders: loading 1`] = `
"<div>
  <h1>
    loading
  </h1>
</div>"
`;
